text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+
+Doing so drastically simplify the entire process, as it means we can effectively drop the entire notion of *cx-protocol*.
+
+## Why is it OK to enforce this requirement?
+
+- We want to migrate most of the functionality to the framework. Making CLI upgrades especially safe.
+- Users tend to upgrade the CLI before upgrading the framework.
+- We sometimes require that even now, and we haven't seen any backlash due to that.
+
+## Why is it NOT OK to enforce this requirement?
+
+- It is a bit intrusive. Even though we currently sometimes do it, we usually don't. And this would mean users get a slightly less smooth upgrade experience.
+
+## How does this affect the design?
+
+As far at testing strategy, it only affects the `test-framework-compatibility.sh` file, which would now look like this:
+
+```bash
+# Framework doesn't support older versions of the CLI, so we expect everything to fail with an appropriate error message.
+run_integration_tests_perm ${LATEST_PUBLISHED_VERSION} "SOURCE ${LATEST_PUBLISHED_VERSION} ${EXPECTED_ERROR}
+```
+
+There are several technical issues we should address of course, but first we need to decide if such an approach is valid.
+
+# Adoption Strategy
+
+Adoption of this strategy should be rather seamless for developers. Like we already mentioned, the changes a developer would need to incorporate, will be guided by failing tests. Hopefully, the process will not require any memorization (or even documentation). 
+
+Having said that, this RFC can be a good reference documentation that provides rational for every step in the process.
+
+# Unresolved questions
+
+There are some scenarios which are still a bit vague with regards to what should be our expected outcome.
+
+## Accidental (In)Compatibility
+
+Imagine a developer pushes some code that changes the `manifest.json`, without bumping the protocol version. Also imagine that this change is compatible with older CLI versions (maybe just an added property), meaning that all integration tests pass. Actually, you don't have to imagine, here is a real life example: [custom docker files support](https://github.com/aws/aws-cdk/pull/5652). Note that this feature only requires a new CLI to support a **new** feature, while all previous functionality remains intact.
+
+With this proposed mechanism, its very likely this code will be released without performing a version bump.
+
+In contrast, imagine the reverse, what if we push code that contains a version bump, making the framework incompatible, while in reality, existing functionality is not broken?
+
+This begs the question, when do we actually want a version bump? On any change to the `manifest.json`? Perhaps only on changes that break the CLI? Maybe we should require CLI upgrades only when detecting usages of specific features?
```

Comment:
We want a version bump for features that the new framework expects from the CLI. Examples could be new context provider lookups, understanding and acting upon new fields in the manifest, etc.

Reply:
Yeah but we are not checking if the feature is actually being used. Ideally we wouldn't require a user to upgrade the CLI if he/she is not even using the feature that actually requires the upgrade.

Reply:
I agree!